Provider/src/main/java/com/spotify/confidence/AndroidLifecycleEventProducer.kt

```diff
@@ -0,0 +1,235 @@
+package com.spotify.confidence
+
+import android.app.Activity
+import android.app.Application
+import android.content.Context
+import android.content.Intent
+import android.content.pm.PackageInfo
+import android.content.pm.PackageManager
+import android.net.ParseException
+import android.net.Uri
+import android.os.Build
+import android.os.Bundle
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.ProcessLifecycleOwner
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+
+class AndroidLifecycleEventProducer(
+    private val application: Application
+) : Application.ActivityLifecycleCallbacks, DefaultLifecycleObserver, EventProducer {
+    private val eventsFlow = MutableSharedFlow<Event>()
+    private val contextFlow = MutableStateFlow<Map<String, ConfidenceValue>>(mapOf())
+    private val sharedPreferences by lazy {
+        application.getSharedPreferences("CONFIDENCE_EVENTS", Context.MODE_PRIVATE)
+    }
+    private val packageInfo: PackageInfo?
+    private val lifecycle: Lifecycle
+
+    private val coroutineScope = CoroutineScope(Dispatchers.IO)
+
+    init {
+        val packageManager: PackageManager = application.packageManager
+        packageInfo = try {
+            packageManager.getPackageInfo(application.packageName, 0)
+        } catch (e: PackageManager.NameNotFoundException) {
+            null
+        }
+
+        // setup lifecycle listeners
+        application.registerActivityLifecycleCallbacks(this)
+        lifecycle = ProcessLifecycleOwner.get().lifecycle
+        coroutineScope.launch(Dispatchers.Main) {
+            lifecycle.addObserver(this@AndroidLifecycleEventProducer)
+        }
+    }
+    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {
+        trackDeepLink(activity)
+        activity.trackActivity("activity-created")
+    }
+
+    override fun onCreate(owner: LifecycleOwner) {
+        trackApplicationLifecycleEvents()
```

Comment:
Would it make sense to add a `APP_STARTED` event here? The challenge is to not include this into the context though

Reply:
Never mind, I see we have `APP_LAUNCHED`